### PR TITLE
End user/feature/register

### PR DESCRIPTION
**Pull request by mosayed01**
**demo video:**
https://youtu.be/Jh_skkFxpsw
 
**in this PR:**
  - [x] refactor login ui to reuse composable
  - [x] add register flow
  - [x] add submit register
  - [x] handle errors
  - [x] handle when success register
  - [x] separate validations to use case

### CODE REVIEW

File: client_end_user/shared/src/commonMain/kotlin/data/gateway/remote/UserGateway.kt [hunk 1]
```diff
@@ -1,20 +1,19 @@
...
-import io.ktor.client.call.body
 import io.ktor.client.request.forms.submitForm
-import io.ktor.client.request.header
 import io.ktor.client.request.url
 import io.ktor.http.HttpMethod
 import io.ktor.http.Parameters
-import io.ktor.http.isSuccess
-import io.ktor.http.parameters
 
 class UserGateway(client: HttpClient) : BaseGateway(client), IUserGateway {
```
Comment (AzizaHelmy) [inline client_end_user/shared/src/commonMain/kotlin/data/gateway/remote/UserGateway.kt:18]: as this is the remote gatway, what about to use the name "UserRemoteGateway" for ex?

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreen.kt [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                ) {
+                    Icon(
+                        painter = painterResource(Resources.images.icBack),
+                        contentDescription = null,
+                        tint = Theme.colors.primary,
+                        modifier = Modifier.size(16.dp)
+                    )
+                }
+            }
+            HeadFirstCard(
```
Comment (AzizaHelmy) [inline client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreen.kt:71]: HeadFirstCard isn't a clean code name "DD

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenModel.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    // endregion
+
+    // region private methods
+    private fun tryCatch(block: () -> Unit) {
+        try {
+            block()
+        } catch (e: AuthorizationException.InvalidUsernameException) {
+            updateState {
+                it.copy(
+                    usernameErrorMsg = e.message ?: "Invalid Username",
```
Comment (AzizaHelmy) [inline client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenModel.kt:54]: but how will localize this msg with this way?

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import presentation.base.ErrorState
+import presentation.base.ErrorState.InvalidEmail
+import presentation.base.ErrorState.InvalidFullName
+import presentation.base.ErrorState.InvalidPhone
+import presentation.base.ErrorState.NoInternet
+import presentation.base.ErrorState.RequestFailed
+import presentation.base.ErrorState.UserAlreadyExists
+import presentation.base.ErrorState.WifiDisabled
+
+class RegistrationSubmitScreenModel(
```
Comment (AzizaHelmy) [inline client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt:20]: as i remeber from uncle Bareq's review he asked us to try another way to pass the data between screens,
check "RestaurantInformationScreenModel" in resturant app to check this way and let's descuss

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                onError = ::onError
+            )
+        }
+    }
+
+    private fun onRegistrationSuccess(success: Boolean) {
+        if (success) {
+            sendNewEffect(RegistrationSubmitScreenEffect.NavigateToLoginScreen)
+        } else {
+            showSnackbar("Opps, Registration not completed")
```
Comment (AzizaHelmy) [inline client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt:63]: avoid hard strings

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitUIState.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+data class RegistrationSubmitUIState(
+    val fullName: String = "",
+    val email: String = "",
+    val phone: String = "",
+    val fullErrorMsg: String = "",
+    val emailErrorMsg: String = "",
+    val phoneErrorMsg: String = "",
+    val snackbarMessage: String = "",
+    val showSnackbar: Boolean = false,
+    val isFullError: Boolean = false,
```
Comment (AzizaHelmy) [inline client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitUIState.kt:12]: i think "isFullNameError" is more readable

File: client_end_user/shared/src/commonMain/kotlin/resources/DrawableResources.kt [hunk 7]
```diff
@@ -26,6 +26,7 @@ data class DrawableResources(
     val cart: String = "img_cart.png",
     val taxiOnTheWay: String = "taxi_on_the_way.png",
     val orderOnTheWay: String = "order_on_the_way.png",
+    val icBack: String = "ic_back.xml",
```
Comment (AzizaHelmy) [inline client_end_user/shared/src/commonMain/kotlin/resources/DrawableResources.kt:29]: i think it's better to use "iconBack" insted

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt (hunk 4 above)
Comment (mosayed01) [reply client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt:20]: https://github.com/TheChance101/beep-beep/blob/develop/client_restaurant/shared/src/commonMain/kotlin/presentation/information/RestaurantInformationScreen.kt

Same way I see

Comment (AzizaHelmy) [reply client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt:20]: i didn't get the meaning you means, but  look at the view model here :
https://github.com/TheChance101/beep-beep/blob/develop/client_restaurant/shared/src/commonMain/kotlin/presentation/information/RestaurantInformationScreenModel.kt
and eng Alhams use "  getRestaurantInfo(get())" to get the args and avoid to pass it in the constructor of view model like the comment of the review was

Comment (AzizaHelmy) [reply client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt:20]: and this in the screen itself :
" initScreen(getScreenModel<RestaurantInformationScreenModel> { parametersOf(id) })"

File: client_end_user/shared/src/commonMain/kotlin/di/screenModelsModule.kt [hunk 8]
```diff
@@ -2,12 +2,18 @@ package di
...
+import presentation.auth.signup.registrationSubmit.RegistrationSubmitScreenModel
 import presentation.home.HomeScreenModel
 import presentation.main.MainScreenModel
 
 val screenModelsModule = module {
     factoryOf(::LoginScreenModel)
     factoryOf(::HomeScreenModel)
     factoryOf(::MainScreenModel)
+    factoryOf(::RegistrationScreenModel)
+    factory { (username: String, password: String) ->
```
Comment (ahmednasserzaza) [inline client_end_user/shared/src/commonMain/kotlin/di/screenModelsModule.kt:16]: you can use factoryOf and pass the parameters in screen

**Review by ahmednasserzaza — APPROVED** [review]
Good work 👏

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt (hunk 4 above)
Comment (mosayed01) [reply client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt:20]: I tried but it's not working with 2 params

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenModel.kt (hunk 3 above)
Comment (mosayed01) [reply client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenModel.kt:54]: here is local validation not server side validation

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreen.kt (hunk 2 above)
Comment (mosayed01) [reply client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreen.kt:71]: in another branch cause there is branches depend on this composable function

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt (hunk 5 above)
Comment (mosayed01) [reply client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt:63]: we will in another branch

**Review by AzizaHelmy — APPROVED** [review]
great work 👏
what about the way of passing args to the view model??
i'm so intersted to know the right way

**Pull request merged**